File: c/hard/leetcode4.c
//
// Created by ywh on 2020/7/22.
//

#include <stdlib.h>

/**
 * 求两个有序数组的中位数
 *
 * @param nums1
 * @param nums1Size
 * @param nums2
 * @param nums2Size
 * @return
 */
double findMedianSortedArrays(int *nums1, int nums1Size, int *nums2, int nums2Size) {
    int idx1 = 0, idx2 = 0, idx = 0, half = (nums1Size + nums2Size) / 2 + 1;
    int *nums = malloc(sizeof(int) * half);
    while (idx < half) {
        int n = 0;
        if (idx1 < nums1Size && idx2 < nums2Size) {
            n = (nums1[idx1] < nums2[idx2]) ? nums1[idx1++] : nums2[idx2++];
        } else if (idx1 < nums1Size) {
            n = nums1[idx1++];
        } else if (idx2 < nums2Size) {
            n = nums2[idx2++];
        }
        nums[idx++] = n;
    }
    if ((nums1Size + nums2Size) % 2) {
        return nums[idx - 1];
    } else {
        return (nums[idx - 1] + nums[idx - 2]) / 2.0;
    }
}

File: go/src/problem/leetcode4.go
package leetcode

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    k, n := 0, (len(nums1)+len(nums2))/2+1
    nums := make([]int, n)
    for i, j := 0, 0; k < n; k++ {
        if i < len(nums1) && j < len(nums2) {
            if nums1[i] < nums2[j] {
                nums[k] = nums1[i]
                i++
            } else {
                nums[k] = nums2[j]
                j++
            }
        } else if i < len(nums1) {
            nums[k] = nums1[i]
            i++
        } else if j < len(nums2) {
            nums[k] = nums2[j]
            j++
        }
    }
    if (len(nums1)+len(nums2))%2 == 1 {
        return float64(nums[k-1])
    } else {
        return float64(nums[k-1]+nums[k-2]) / 2
    }
}
